core/src/main/scala/io/qbeast/spark/index/SparkRevisionFactory.scala

```diff
@@ -58,77 +56,265 @@ object SparkRevisionFactory extends RevisionFactory {
+          // If the existing Transformer is an EmptyTransformer
+          Some(newSpec.toTransformer(schema))
+        case (newSpec @ ColumnToIndex(_, Some(newType)), ColumnToIndex(_, Some(oldType)))
+            if newType != oldType => // If the transformerType has changed
+          Some(newSpec.toTransformer(schema))
+        case _ => None
+      }
+      .toIndexedSeq
+  }
+
+  /**
+   * Compute the updated transformers. If a new transformer is created, it will replace the
+   * existing one.
+   * @param transformers
+   *   the existing transformers
+   * @param transformerChanges
+   *   the transformer changes created by user input
+   * @return
+   */
+  private[index] def computeUpdatedTransformers(
+      transformers: IISeq[Transformer],
+      transformerChanges: IISeq[Option[Transformer]]): IISeq[Transformer] = {
+    transformers.zip(transformerChanges).map {
+      case (_, Some(newT)) => newT
+      case (t, None) => t
     }
+  }
 
-    val revision =
-      Revision.firstRevision(qtableID, desiredCubeSize, transformers, transformations)
+  /**
+   * Analyze and extract statistics for the indexing columns
+   * @param data
+   *   the data to analyze
+   * @param transformers
+   *   the columns to analyze
+   */
+  private[index] def getDataFrameStats(data: DataFrame, transformers: IISeq[Transformer]): Row = {
+    val columnsStatsExpr = transformers.map(_.stats).flatMap(_.statsSqlPredicates)
+    logInfo("Computing statistics: " + columnsStatsExpr.mkString(", "))
+    val row = data.selectExpr(columnsStatsExpr ++ Seq("count(1) AS count"): _*).first()
+    logInfo("Computed statistics: " + row.mkString(", "))
+    row
+  }
 
-    // When all indexing columns have been provided with a boundary, update the RevisionID
-    // to 1 to avoid using the StagingRevisionID(0). It is possible for this RevisionID to
-    // to be later updated to 2 if the actual column boundaries are larger that than the
-    // provided values. In this case, we will have Revision 2 instead of Revision 1.
-    if (shouldCreateNewSpace) revision.copy(revisionID = 1)
-    else revision
+  /**
+   * Compute the cube size changes for the current operation
+   * @param revision
+   *   the existing revision
+   * @param options
+   *   the QbeastOptions
+   */
+  private[index] def computeCubeSizeChanges(
+      revision: Revision,
+      options: QbeastOptions): Option[Int] = {
+    if (revision.desiredCubeSize != options.cubeSize) Some(options.cubeSize)
+    else None
   }
 
-  override def createNextRevision(
-      qtableID: QTableID,
-      schema: StructType,
+  /**
+   * Compute the transformation changes for the current operation. The process of creating
+   * transformation changes is as follows:
+   *   1. Compute the Transformations from the input columnStats(if available) and the input
+   *      DataFrame statistics. 2. Check if the new transformations supersede the existing ones.
```

Comment:
We need to pick a better formatting for these comments...

Reply:
Done